apps/axistream/src/axi_stream_reader.hh

```diff
@@ -0,0 +1,168 @@
+        return T::from_raw((read_raw<to_read / util::BITS_PER_BYTE>(), H::to_raw(header)));
+    }
+
+    /*
+     * Given some message T which starts with header H, reads all non-header bytes
+     * of T from the stream, then concatenates the new bytes with the given header
+     * to return a complate T.
+     *
+     * This overload works for types T which are the same size as the header - in other words
+     * the type consists of only a header.
+     */
+    template <typename T, typename H>
+    typename std::enable_if<T::width == H::width, T>::type read_after(const H& header) {
+        HLS_PRAGMA(HLS inline);
+        return T::from_raw(H::to_raw(header));
+    }
+
+    /*
+     * Reads the specified number of byes from the stream.
+     */
+    template <int MsgWidth>
+    util::Bytes<MsgWidth> read_raw() {
+        HLS_PRAGMA(HLS inline);
+        // The size of the working buffer.
+        // This handles the worst-case where we have DataWidth - 1 bytes in buf_,
+        // and so the last byte of the message data will be at byte index DataWidth - 1 + MsgWidth - 1
+        // (we will need to store DataWidth - 1 + MsgWidth bytes).
+        // The work size, in the last data beat read, the high byte will not be part of this message,
+        // but we still must store it in the work buffer.
+        // Thus the work buffer neesd to be DataWidth + MsgWidth bytes.
+        const int work_size = util::BITS_PER_BYTE * (DataWidth + MsgWidth);
+        // Maximum number of beats required to get the whole message = ceil(MsgWidth / DataWidth)
+        // It is possible the last beat isn't needed depending on how much is stored in buf_
+        // from the last read
+        const int max_beats = (MsgWidth + DataWidth - 1) / DataWidth;
+        // Maximum number of new bytes we might receive minus 1
+        // This is the maximal value received may take as of the end of the second to last
+        // loop iteration, and as such the maximal value received needs to ever hold
+        // (it doesn't matter if it overflows during the last loop iteration, because it will
+        // never be read again).
+        const int max_received = DataWidth * max_beats - 1;
+        // Type which can store the number of bytes to shift the final result by
+        // At most, all the low DataWidth bytes need to be shifted
+        typedef ap_uint<util::unsigned_bit_width<DataWidth * util::BITS_PER_BYTE>::value> Shift;
+
+        // Buf will be zero extended since work is MsgWidth longer than buf_
+        ap_uint<work_size> work = buf_;
+        // Represents the number of bytes we have so far
+        ap_uint<util::unsigned_bit_width<max_received>::value> received = len_;
+        // Represents the index in work where the last byte of the message will be written
+        ap_uint<util::unsigned_bit_width<work_size - 1>::value> last_index = DataWidth - len_ + MsgWidth - 1;
+
+        for (int i = 0; i < max_beats; i++) {
+            HLS_PRAGMA(HLS unroll);
```

Comment:
This unroll is somewhat problematic.  It's somewhat a tradeoff whether this should be a pipelined loop or not.  It depends a bit on the context in which you expect the code to be used.  What examples have you built?

Reply:
I need to test this more. The interesting thing about unrolling this loop is that if you read multiple messages in sequence, and all the read calls are inlined and this loop is unrolled, the offset for each field in the _stream_ is statically known. As for being a pipelined loop, it has an interation latency of 1 cycle anyway.

Reply:
I agree, there is a huge value in making the offsets statically known.  There are actually several places where this comes in.  The simplest one is ensuring that offsets in a header are statically known.  If they are not statically known, then narrowing them down to a small set of options is important (offset is either 64 or 96, but it's not arbitrary).    The second one is ensuring that the  offset within the width of the AXI stream where data starts is constant.  I suspect that unrolling this loop primarily helps with having constant offsets in a header.  It also helps reduce the loop overhead generated by HLS.  (this is more of a limitation in HLS, since max_beats is often very small for interesting packet formats, which means pipelining this loop can have significant performance impacts.

Reply:
It looks like you've focused on the model of accessing one header at a time.  Did you consider the model of operating on a whole packet at a time?

Reply:
I've kind of left that till the next layer - this code just reads a message (a single header) from a stream. Then if someone wants a whole packet, they'd call read to get a bunch of headers:
``` C++
ethernet_header e = reader.read<ethernet_header>();
ip_header ip = reader.read<ip_header>();
// ...
```

If after they read the headers, they want the data, then they could just call `read_raw<8>`, for example, to get 8 bytes at a time. And if they want to store the data in a RAM, they do that themselves:
```C++
char data[100];
for (int i = 0; i < len; i++) {
   data[i] = reader.read_raw<1>();
}
```

But if they want to process the data and send it back out at the same time, they just don't store it, but still loop through it.
Then the user does whatever they want with the data later.

Reply:
Also, this unroll was based on: https://github.com/Xilinx/HLS_packet_processing/blob/efbd9996fbc5568925e2d851c64f5d2106657d51/apps/common/ip.hpp#L2849-L2851

Reply:
The code to get the ending condition is tricky while writing it in a width-parameterizable way.  This is one reason why I like the 'packet' model for writing the code.